src/simsopt/geo/curve.py

```diff
@@ -441,21 +477,72 @@ def __init__(self, quadpoints, gamma_pure, **kwargs):
         self.dgammadashdashdash_by_dcoeff_jax = jit(jacfwd(self.gammadashdashdash_jax))
         self.dgammadashdashdash_by_dcoeff_vjp_jax = jit(lambda x, v: vjp(self.gammadashdashdash_jax, x)[1](v)[0])
 
+        self.centroid_jax = jit(lambda x, v: centroid_pure(x, v))
+        self.dcentroid_dgamma = jit(jacfwd(self.centroid_jax, argnums=0))
+        self.dcentroid_dgammadash = jit(jacfwd(self.centroid_jax, argnums=1))
+        self.kappa_pure = kappa_pure
+        self.kappa_jax = jit(lambda x, v: kappa_pure(x, v))
+        self.kappa_impl_jax = jit(lambda x, v: kappa_pure(x, v))
+        self.frenet_frame_jax = jit(lambda x: self.frenet_frame_pure(x))
+        self.incremental_arclength_jax = jit(lambda x: self.incremental_arclength_pure(x))
         self.dkappa_by_dcoeff_vjp_jax = jit(lambda x, v: vjp(lambda d: kappa_pure(self.gammadash_jax(d), self.gammadashdash_jax(d)), x)[1](v)[0])
-
         self.dtorsion_by_dcoeff_vjp_jax = jit(lambda x, v: vjp(lambda d: torsion_pure(self.gammadash_jax(d), self.gammadashdash_jax(d), self.gammadashdashdash_jax(d)), x)[1](v)[0])
 
     def set_dofs(self, dofs):
+        """
+        This function sets the dofs of the curve.
+        """
         self.local_x = dofs
         sopp.Curve.set_dofs(self, dofs)
 
     def gamma_impl(self, gamma, quadpoints):
         r"""
         This function returns the x,y,z coordinates of the curve :math:`\Gamma`.
         """
-
         gamma[:, :] = self.gamma_impl_jax(self.get_dofs(), quadpoints)
 
+    def incremental_arclength_pure(self, dofs):
+        """
+        This function returns the incremental arclength of the curve.
+        """
+        gammadash = self.gammadash_jax(dofs)
+        return jnp.linalg.norm(gammadash, axis=1)
+
+    def incremental_arclength(self):
+        """
+        This function returns the incremental arclength of the curve.
+        """
+        return self.incremental_arclength_jax(self.get_dofs())
+
+    # @jit
+    def frenet_frame_pure(self, dofs):
+        r"""
+        This function returns the Frenet frame, :math:`(\mathbf{t}, \mathbf{n}, \mathbf{b})`,
+        associated to the curve.
+        """
+        gammadash = self.gammadash_jax(dofs)
+        gammadashdash = self.gammadashdash_jax(dofs)
+        l = self.incremental_arclength()
+        def norm(a): return jnp.linalg.norm(a, axis=1)
+        def inner(a, b): return jnp.sum(a*b, axis=1)
+        N = len(self.quadpoints)
+        t, n, b = (jnp.zeros((N, 3)), jnp.zeros((N, 3)), jnp.zeros((N, 3)))
+        t = (1./l[:, None]) * gammadash
+
+        tdash = (1./l[:, None])**2 * (l[:, None] * gammadashdash
+                                      - (inner(gammadash, gammadashdash)/l)[:, None] * gammadash
+                                      )
+        n = (1./norm(tdash))[:, None] * tdash
+        b = jnp.cross(t, n, axis=1)
+        return t, n, b
+
+    def frenet_frame(self):
```

Comment:
hm, since JaxCurve is a Curve, doesn't `frenet_frame` already exist for this class?